File: apis/volumes.py
from flask import Blueprint, request, jsonify
import docker

volumes_blueprint = Blueprint('volume', __name__)
client = docker.from_env()

@volumes_blueprint.route('/volumes', methods=['GET'])
def list_volumes():
    volumes = client.volumes.list()
    volume_list = [{
        'name': volume.name,
        'driver': volume.attrs.get('Driver', 'unknown'),
        'mountpoint': volume.attrs.get('Mountpoint', 'unknown'),
        'labels': volume.attrs.get('Labels', {}),
        'scope': volume.attrs.get('Scope', 'unknown'),
        'options': volume.attrs.get('Options', {}),
    } for volume in volumes]
    return jsonify(volume_list)

# Volume Management Endpoints
@volumes_blueprint.route('/volumes/create', methods=['POST'])
def create_volume():
    data = request.json
    name = data.get('name')
    driver = data.get('driver', 'local')
    options = data.get('options', {})

    volume = client.volumes.create(name=name, driver=driver, driver_opts=options)
    return jsonify({'name': volume.name, 'driver': volume.driver, 'mountpoint': volume.attrs['Mountpoint']}), 201

@volumes_blueprint.route('/volumes/remove/<volume_name>', methods=['DELETE'])
def remove_volume(volume_name):
    try:
        client.volumes.get(volume_name).remove()
        return '', 204
    except docker.errors.NotFound:
        return 'Volume not found', 404

@volumes_blueprint.route('/volumes/update/<volume_name>', methods=['POST'])
def update_volume(volume_name):
    data = request.json
    options = data.get('options', {})

    try:
        volume = client.volumes.get(volume_name)
        volume.driver_opts.update(options)
        return jsonify({'name': volume.name, 'driver': volume.driver, 'mountpoint': volume.attrs['Mountpoint']})
    except docker.errors.NotFound:
        return 'Volume not found', 404

# External Mount Management Endpoints
@volumes_blueprint.route('/mounts/create', methods=['POST'])
def create_mount():
    data = request.json
    mount_type = data.get('type')
    mount_point = data.get('mount_point')
    source = data.get('source')
    options = data.get('options', {})

    # For demonstration, this example only handles 'smb' type mount
    if mount_type == 'smb':
        mount_command = f'sudo mount -t cifs {source} {mount_point} -o ' + ','.join([f'{key}={value}' for key, value in options.items()])
        # Execute the mount command
        import subprocess
        subprocess.run(mount_command, shell=True, check=True)
        return jsonify({'mount_point': mount_point, 'source': source, 'options': options}), 201
    else:
        return 'Mount type not supported', 400

@volumes_blueprint.route('/mounts/remove/<mount_id>', methods=['DELETE'])
def remove_mount(mount_id):
    # Example assumes that 'mount_id' is a mount point
    umount_command = f'sudo umount {mount_id}'
    import subprocess
    subprocess.run(umount_command, shell=True, check=True)
    return '', 204

@volumes_blueprint.route('/mounts/update/<mount_id>', methods=['POST'])
def update_mount(mount_id):
    data = request.json
    new_mount_point = data.get('mount_point')
    new_source = data.get('source')
    new_options = data.get('options', {})

    # Example assumes the mount was originally handled by SMB
    umount_command = f'sudo umount {mount_id}'
    import subprocess
    subprocess.run(umount_command, shell=True, check=True)

    mount_command = f'sudo mount -t cifs {new_source} {new_mount_point} -o ' + ','.join([f'{key}={value}' for key, value in new_options.items()])
    subprocess.run(mount_command, shell=True, check=True)

    return jsonify({'mount_point': new_mount_point, 'source': new_source, 'options': new_options}), 200

File: frontend/dev/api/volumes.ts
const API_BASE_URL = 'http://localhost:5000'; // API temel URL'si

// Tüm volume'leri listele
export const listVolumes = async (): Promise<any[]> => {
  const response = await fetch(`${API_BASE_URL}/volumes`);
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Yeni bir volume oluştur
export const createVolume = async (data: any): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/volumes/create`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Bir volume'ü sil
export const removeVolume = async (volumeName: string): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/volumes/remove/${volumeName}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Bir volume'ü güncelle
export const updateVolume = async (volumeName: string, data: any): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/volumes/update/${volumeName}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Yeni bir mount oluştur
export const createMount = async (data: any): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/mounts/create`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Bir mount'ı sil
export const removeMount = async (mountId: string): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/mounts/remove/${mountId}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Bir mount'ı güncelle
export const updateMount = async (mountId: string, data: any): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/mounts/update/${mountId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};
